nexus/db-queries/src/db/datastore/image.rs

```diff
@@ -17,50 +18,149 @@ use crate::db::error::public_error_from_diesel_pool;
 use crate::db::error::ErrorHandler;
 use crate::db::model::Image;
 use crate::db::model::Project;
+use crate::db::model::ProjectImage;
+use crate::db::model::Silo;
+use crate::db::model::SiloImage;
 use crate::db::pagination::paginated;
 
 use async_bb8_diesel::AsyncRunQueryDsl;
 
 use super::DataStore;
 
 impl DataStore {
-    pub async fn image_list(
+    pub async fn project_image_list(
         &self,
         opctx: &OpContext,
         authz_project: &authz::Project,
+        include_silo_images: bool,
         pagparams: &PaginatedBy<'_>,
     ) -> ListResultVec<Image> {
         opctx.authorize(authz::Action::ListChildren, authz_project).await?;
 
         use db::schema::image::dsl;
+        use db::schema::project_image::dsl as project_dsl;
+        match include_silo_images {
+            true => match pagparams {
+                PaginatedBy::Id(pagparams) => {
+                    paginated(dsl::image, dsl::id, &pagparams)
+                }
+                PaginatedBy::Name(pagparams) => paginated(
+                    dsl::image,
+                    dsl::name,
+                    &pagparams.map_name(|n| Name::ref_cast(n)),
+                ),
+            }
+            .filter(dsl::time_deleted.is_null())
+            .filter(
+                dsl::project_id
+                    .is_null()
+                    .or(dsl::project_id.eq(authz_project.id())),
+            )
+            .select(Image::as_select())
+            .load_async::<Image>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            }),
+            false => match pagparams {
+                PaginatedBy::Id(pagparams) => paginated(
+                    project_dsl::project_image,
+                    project_dsl::id,
+                    &pagparams,
+                ),
+                PaginatedBy::Name(pagparams) => paginated(
+                    project_dsl::project_image,
+                    project_dsl::name,
+                    &pagparams.map_name(|n| Name::ref_cast(n)),
+                ),
+            }
+            .filter(project_dsl::time_deleted.is_null())
+            .filter(project_dsl::project_id.eq(authz_project.id()))
+            .select(ProjectImage::as_select())
+            .load_async::<ProjectImage>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+            .map(|v| v.into_iter().map(|v| v.into()).collect()),
```

Comment:
Note that we're still returning the generic `Image` type here so we have to map over the results and convert them in this case.